File: minecraft/src/main/java/net/exohayvan/dissolver_enhanced/common/analytics/PostHogErrorReporter.java
package net.exohayvan.dissolver_enhanced.common.analytics;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HexFormat;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.function.BiConsumer;
import java.util.function.Supplier;

public final class PostHogErrorReporter implements AutoCloseable {
    private static final String EXCEPTION_EVENT = "$exception";
    private static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(5);

    private final URI endpoint;
    private final String token;
    private final String distinctId;
    private final String inAppPackagePrefix;
    private final Supplier<Map<String, Object>> baseProperties;
    private final HttpClient httpClient;
    private final ExecutorService executor;
    private final BiConsumer<String, Throwable> warningLogger;

    public PostHogErrorReporter(
        String endpoint,
        String token,
        String distinctId,
        String inAppPackagePrefix,
        Supplier<Map<String, Object>> baseProperties,
        BiConsumer<String, Throwable> warningLogger
    ) {
        this.endpoint = URI.create(Objects.requireNonNull(endpoint, "endpoint"));
        this.token = Objects.requireNonNull(token, "token");
        this.distinctId = Objects.requireNonNull(distinctId, "distinctId");
        this.inAppPackagePrefix = Objects.requireNonNull(inAppPackagePrefix, "inAppPackagePrefix");
        this.baseProperties = baseProperties;
        this.warningLogger = warningLogger;
        this.executor = Executors.newSingleThreadExecutor(task -> {
            Thread thread = new Thread(task, "dissolver-enhanced-posthog-errors");
            thread.setDaemon(true);
            return thread;
        });
        this.httpClient = HttpClient.newBuilder()
            .connectTimeout(REQUEST_TIMEOUT)
            .executor(executor)
            .build();
    }

    public CompletableFuture<Void> capture(Throwable throwable, boolean handled) {
        if (!hasInAppFrame(throwable)) {
            return CompletableFuture.completedFuture(null);
        }

        Map<String, Object> properties = new LinkedHashMap<>();
        if (baseProperties != null) {
            properties.putAll(baseProperties.get());
        }
        properties.put("distinct_id", distinctId);
        properties.put("$exception_list", List.of(exceptionObject(throwable, handled, inAppPackagePrefix)));
        properties.put("$exception_fingerprint", fingerprint(throwable));
        properties.put("$exception_message", exceptionMessage(throwable));
        properties.put("$exception_type", throwable.getClass().getName());
        properties.put("$exception_stack_trace_raw", rawStackTrace(throwable));
        properties.put("exception_handled", handled);
        properties.put("exception_causes", exceptionCauses(throwable));

        Map<String, Object> body = new LinkedHashMap<>();
        body.put("token", token);
        body.put("event", EXCEPTION_EVENT);
        body.put("properties", properties);
        body.put("timestamp", Instant.now().toString());

        HttpRequest request = HttpRequest.newBuilder(endpoint)
            .timeout(REQUEST_TIMEOUT)
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(PostHogJson.toJson(body)))
            .build();

        return httpClient.sendAsync(request, HttpResponse.BodyHandlers.discarding())
            .thenAccept(response -> {
                int statusCode = response.statusCode();
                if (statusCode < 200 || statusCode >= 300) {
                    logWarning("PostHog exception capture failed with HTTP " + statusCode + ".", null);
                }
            })
            .exceptionally(exception -> {
                logWarning("PostHog exception capture failed.", exception);
                return null;
            });
    }

    @Override
    public void close() {
        executor.shutdown();
    }

    private boolean hasInAppFrame(Throwable throwable) {
        Throwable current = throwable;
        while (current != null) {
            for (StackTraceElement element : current.getStackTrace()) {
                if (isInAppFrame(element, inAppPackagePrefix)) {
                    return true;
                }
            }
            current = current.getCause();
        }

        return false;
    }

    private static Map<String, Object> exceptionObject(Throwable throwable, boolean handled, String inAppPackagePrefix) {
        Map<String, Object> exception = new LinkedHashMap<>();
        exception.put("type", throwable.getClass().getName());
        exception.put("value", exceptionMessage(throwable));
        exception.put("mechanism", mechanism(handled));
        exception.put("stacktrace", stacktrace(throwable, inAppPackagePrefix));
        return exception;
    }

    private static Map<String, Object> mechanism(boolean handled) {
        Map<String, Object> mechanism = new LinkedHashMap<>();
        mechanism.put("handled", handled);
        mechanism.put("synthetic", false);
        return mechanism;
    }

    private static Map<String, Object> stacktrace(Throwable throwable, String inAppPackagePrefix) {
        Map<String, Object> stacktrace = new LinkedHashMap<>();
        stacktrace.put("type", "raw");
        stacktrace.put("frames", frames(throwable, inAppPackagePrefix));
        return stacktrace;
    }

    private static List<Map<String, Object>> frames(Throwable throwable, String inAppPackagePrefix) {
        List<Map<String, Object>> frames = new ArrayList<>();
        for (StackTraceElement element : throwable.getStackTrace()) {
            Map<String, Object> frame = new LinkedHashMap<>();
            frame.put("platform", "custom");
            frame.put("lang", "java");
            frame.put("function", element.getMethodName());
            frame.put("filename", element.getFileName());
            frame.put("lineno", element.getLineNumber());
            frame.put("module", element.getClassName());
            frame.put("resolved", true);
            frame.put("in_app", isInAppFrame(element, inAppPackagePrefix));
            frames.add(frame);
        }
        return frames;
    }

    private static String exceptionMessage(Throwable throwable) {
        return throwable.getMessage() == null ? throwable.toString() : throwable.getMessage();
    }

    private static String rawStackTrace(Throwable throwable) {
        StringWriter writer = new StringWriter();
        throwable.printStackTrace(new PrintWriter(writer));
        return writer.toString();
    }

    private static List<Map<String, Object>> exceptionCauses(Throwable throwable) {
        List<Map<String, Object>> causes = new ArrayList<>();
        Throwable current = throwable.getCause();
        while (current != null) {
            Map<String, Object> cause = new LinkedHashMap<>();
            cause.put("type", current.getClass().getName());
            cause.put("message", exceptionMessage(current));
            causes.add(cause);
            current = current.getCause();
        }
        return causes;
    }

    private static boolean isInAppFrame(StackTraceElement element, String inAppPackagePrefix) {
        return element.getClassName().startsWith(inAppPackagePrefix);
    }

    private static String fingerprint(Throwable throwable) {
        StackTraceElement[] stackTrace = throwable.getStackTrace();
        String topFrame = stackTrace.length == 0 ? "unknown" : stackTrace[0].toString();
        String rawFingerprint = throwable.getClass().getName() + ":" + topFrame;
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            return HexFormat.of().formatHex(digest.digest(rawFingerprint.getBytes(StandardCharsets.UTF_8)));
        } catch (NoSuchAlgorithmException e) {
            return Integer.toHexString(rawFingerprint.hashCode());
        }
    }

    private void logWarning(String message, Throwable throwable) {
        if (warningLogger != null) {
            warningLogger.accept(message, throwable);
        }
    }
}

File: stardewvalley/src/Analytics/PostHogErrorReporter.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Diagnostics;
using System.Reflection;

namespace DissolverEnhanced.StardewValley.Common.Analytics;

public sealed class PostHogErrorReporter : IDisposable
{
    private const string ExceptionEvent = "$exception";
    private static readonly TimeSpan RequestTimeout = TimeSpan.FromSeconds(5);
    private static readonly TimeSpan FailureLogCooldown = TimeSpan.FromMinutes(5);
    private readonly HttpClient httpClient;
    private readonly string endpoint;
    private readonly string token;
    private readonly string distinctId;
    private readonly string inAppPackagePrefix;
    private readonly Func<IReadOnlyDictionary<string, object?>> baseProperties;
    private readonly Action<string, Exception?> warningLogger;
    private DateTimeOffset nextFailureLogAt = DateTimeOffset.MinValue;

    public PostHogErrorReporter(
        string endpoint,
        string token,
        string distinctId,
        string inAppPackagePrefix,
        Func<IReadOnlyDictionary<string, object?>> baseProperties,
        Action<string, Exception?> warningLogger
    )
    {
        this.endpoint = endpoint;
        this.token = token;
        this.distinctId = distinctId;
        this.inAppPackagePrefix = inAppPackagePrefix;
        this.baseProperties = baseProperties;
        this.warningLogger = warningLogger;
        httpClient = new HttpClient
        {
            Timeout = RequestTimeout
        };
    }

    public Task CaptureAsync(Exception exception, bool handled)
    {
        if (!HasInAppFrame(exception))
        {
            return Task.CompletedTask;
        }

        Dictionary<string, object?> properties = new(baseProperties())
        {
            ["distinct_id"] = distinctId,
            ["$exception_list"] = new[] { ExceptionObject(exception, handled) },
            ["$exception_fingerprint"] = Fingerprint(exception),
            ["$exception_level"] = "error",
            ["$exception_message"] = ExceptionMessage(exception),
            ["$exception_type"] = exception.GetType().FullName,
            ["$exception_stack_trace_raw"] = exception.ToString(),
            ["exception_handled"] = handled,
            ["exception_message"] = ExceptionMessage(exception),
            ["exception_type"] = exception.GetType().FullName,
            ["exception_top_frame"] = TopFrame(exception),
            ["exception_in_app_top_frame"] = InAppTopFrame(exception),
            ["exception_stack_trace"] = exception.ToString(),
            ["exception_causes"] = ExceptionCauses(exception)
        };

        Dictionary<string, object?> body = new()
        {
            ["token"] = token,
            ["event"] = ExceptionEvent,
            ["properties"] = properties,
            ["timestamp"] = DateTimeOffset.UtcNow.ToString("O")
        };

        return SendAsync(body);
    }

    private async Task SendAsync(IReadOnlyDictionary<string, object?> body)
    {
        try
        {
            string json = JsonSerializer.Serialize(body);
            using StringContent content = new(json, Encoding.UTF8, "application/json");
            using HttpResponseMessage response = await httpClient.PostAsync(endpoint, content).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                warningLogger($"PostHog exception capture failed with HTTP {(int)response.StatusCode}.", null);
            }
        }
        catch (Exception exception)
        {
            LogFailure("PostHog exception capture failed", exception);
        }
    }

    private void LogFailure(string message, Exception exception)
    {
        DateTimeOffset now = DateTimeOffset.UtcNow;
        if (now < nextFailureLogAt)
        {
            return;
        }

        nextFailureLogAt = now.Add(FailureLogCooldown);
        warningLogger($"{message}: {exception.GetType().Name}: {exception.Message}", null);
    }

    private bool HasInAppFrame(Exception exception)
    {
        Exception? current = exception;
        while (current != null)
        {
            if ((current.StackTrace ?? "").Contains(inAppPackagePrefix, StringComparison.Ordinal))
            {
                return true;
            }

            current = current.InnerException;
        }

        return false;
    }

    private Dictionary<string, object?> ExceptionObject(Exception exception, bool handled)
    {
        return new Dictionary<string, object?>
        {
            ["type"] = exception.GetType().FullName,
            ["value"] = ExceptionMessage(exception),
            ["mechanism"] = new Dictionary<string, object?>
            {
                ["handled"] = handled,
                ["type"] = handled ? "generic" : "on_error",
                ["synthetic"] = false
            },
            ["stacktrace"] = new Dictionary<string, object?>
            {
                ["frames"] = Frames(exception)
            }
        };
    }

    private static string ExceptionMessage(Exception exception)
    {
        return string.IsNullOrWhiteSpace(exception.Message) ? exception.ToString() : exception.Message;
    }

    private static List<Dictionary<string, object?>> ExceptionCauses(Exception exception)
    {
        List<Dictionary<string, object?>> causes = new();
        Exception? current = exception.InnerException;
        while (current != null)
        {
            causes.Add(new Dictionary<string, object?>
            {
                ["type"] = current.GetType().FullName,
                ["message"] = ExceptionMessage(current)
            });
            current = current.InnerException;
        }

        return causes;
    }

    private List<Dictionary<string, object?>> Frames(Exception exception)
    {
        List<Dictionary<string, object?>> frames = new();
        foreach (StackFrame frame in new StackTrace(exception, true).GetFrames() ?? Array.Empty<StackFrame>())
        {
            MethodBase? method = frame.GetMethod();
            Type? declaringType = method?.DeclaringType;
            string module = declaringType?.FullName ?? "unknown";

            frames.Add(new Dictionary<string, object?>
            {
                ["platform"] = "custom",
                ["lang"] = "csharp",
                ["function"] = method?.Name ?? "unknown",
                ["filename"] = frame.GetFileName(),
                ["lineno"] = frame.GetFileLineNumber(),
                ["module"] = module,
                ["resolved"] = true,
                ["in_app"] = module.StartsWith(inAppPackagePrefix, StringComparison.Ordinal)
            });
        }

        return frames;
    }

    private static string TopFrame(Exception exception)
    {
        StackFrame? frame = (new StackTrace(exception, true).GetFrames() ?? Array.Empty<StackFrame>()).FirstOrDefault();
        return FrameName(frame);
    }

    private string InAppTopFrame(Exception exception)
    {
        foreach (StackFrame frame in new StackTrace(exception, true).GetFrames() ?? Array.Empty<StackFrame>())
        {
            MethodBase? method = frame.GetMethod();
            string module = method?.DeclaringType?.FullName ?? "";
            if (module.StartsWith(inAppPackagePrefix, StringComparison.Ordinal))
            {
                return FrameName(frame);
            }
        }

        return "unknown";
    }

    private static string FrameName(StackFrame? frame)
    {
        if (frame == null)
        {
            return "unknown";
        }

        MethodBase? method = frame.GetMethod();
        string module = method?.DeclaringType?.FullName ?? "unknown";
        string function = method?.Name ?? "unknown";
        int line = frame.GetFileLineNumber();
        return line > 0 ? $"{module}.{function}:{line}" : $"{module}.{function}";
    }

    private static string Fingerprint(Exception exception)
    {
        string rawFingerprint = exception.GetType().FullName + ":" + (exception.StackTrace?.Split('\n').FirstOrDefault() ?? "unknown");
        byte[] digest = SHA256.HashData(Encoding.UTF8.GetBytes(rawFingerprint));
        return Convert.ToHexString(digest).ToLowerInvariant();
    }

    public void Dispose()
    {
        httpClient.Dispose();
    }
}
